release-tools/README.md

```diff
@@ -0,0 +1,51 @@
+# [csi-release-tools](https://github.com/kubernetes-csi/csi-release-tools)
+
+These build and test rules can be shared between different Go projects
+without modifications. Customization for the different projects happen
+in the top-level Makefile.
+
+The rules include support for building and pushing Docker images, with
+the following features:
+ - one or more command and image per project
+ - push canary and/or tagged release images
+ - automatically derive the image tag(s) from repo tags
+ - the source code revision is stored in a "revision" image label
+ - never overwrites an existing release image
+
+Usage
+-----
+
+The expected repository layout is:
+ - `cmd/*/*.go` - source code for each command
+ - `cmd/*/Dockerfile` - docker file for each command or
+   Dockerfile in the root when only building a single command
+ - `Makefile` - includes `release-tools/build.make` and sets
+   configuration variables
+ - `.travis.yml` - a symlink to `release-tools/.travis.yml`
+
+To create a release, tag a certain revision with a name that
+starts with `v`, for example `v1.0.0`, then `make push`
+while that commit is checked out.
+
+It does not matter on which branch that revision exists, i.e. it is
+possible to create releases directly from master. A release branch can
+still be created for maintenance releases later if needed.
+
+Release branches are expected to be named `release-x.y` for releases
+`x.y.z`. Building from such a branch creates `x.y-canary`
+images. Building from master creates the main `canary` image.
+
+Sharing and updating
+--------------------
+
+[`git subtree`](https://github.com/git/git/blob/master/contrib/subtree/git-subtree.txt)
```

Comment:
I'm not familiar with git subtree, so I have a few questions on the general process:
* Do we need to make a PR in every sidecar repo if we want to sync the subtree to the latest?
* Is there protection against changing these files in the sidecar repos instead of the main one?

Reply:
> Do we need to make a PR in every sidecar repo if we want to sync the subtree to the latest?

Yes. In that regard it is the same as vendoring.

> Is there protection against changing these files in the sidecar repos instead of the main one?

No. This is something that needs to be checked during the code review.

Technically, a sidecar repo can have local commits touching the `release-tools` content. Such modifications will even survive a resync with upstream, in contrast to `dep`, which will wipe out all local modifications. But there's a risk of merge conflicts and it would defeat the goal of having shared build rules, so I think we shouldn't allow that.

But it is useful while working on a change:
- try out some change in a sidecar repo
- submit a PR for review and automated testing
- once the PR is green, take the commit from the sidecar repo and publish it as PR against the `csi-release-tools` repo; the third line item in the cheat sheet shows how to do that
- when the upstream PR has been merged, do one of:
  - `git subtree pull` in the sidecar (easy, but leads to a slightly more complex git history)
  - start anew locally, `git subtree pull`, then cherry-pick whatever other local changes were in the PR (if none, then this is also simple; the result will be a clean git history without a trace of having worked on this locally first)

Reply:
> > Is there protection against changing these files in the sidecar repos instead of the main one?
>
> No. This is something that needs to be checked during the code review.

Let me turn that into a "yes". I have added a check script, see the new commits in this PR. As intended, those changes now cause the PR to not pass CI testing:
https://travis-ci.org/kubernetes-csi/external-attacher/builds/483410807

Note that this will force us into using the "start anew locally" approach for PRs. I personally would enable the check.

Reply:
This sounds good to me

Reply:
@msau42 Then let's go for it. Please approve https://github.com/kubernetes-csi/csi-release-tools/pull/3 and I'll update this PR here.

That PR was created with `git subtree push` as explained in the README.md.